libraries/DataFlash/DataFlash.cpp

```diff
@@ -100,6 +105,129 @@ void DataFlash_Class::Init(const struct LogStructure *structures, uint8_t num_ty
+}
+
+/// pretty-print log structures
+/// @note structures MUST be well-formed
+void DataFlash_Class::dump_structures(const struct LogStructure *structures, const uint8_t num_types)
+{
+#if DEBUG_LOG_STRUCTURES
+    for (uint16_t i=0; i<num_types; i++) {
+        const struct LogStructure *structure = &structures[i];
+        ::fprintf(stderr, "%s\n", structure->name);
+        char label[32] = { };
+        uint8_t labeloffset = 0;
+        int8_t fieldnum = 0;
+        for (uint8_t j=0; j<strlen(structure->labels); j++) {
+            char labelchar = structure->labels[j];
+            if (labelchar == '\0') {
+                break;
+            }
+            if (labelchar == ',') {
+                dump_structure_field(structure, label, fieldnum);
+                fieldnum++;
+                labeloffset = 0;
+                memset(label, '\0', 32);
+            } else {
+                label[labeloffset++] = labelchar;
+            }
+        }
+        dump_structure_field(structure, label, fieldnum);
+        ::fprintf(stderr, "\n"); // just add a CR to the output
+    }
+#endif
+}
+
+bool DataFlash_Class::validate_structures(const struct LogStructure *structures, const uint8_t num_types)
+{
+    Debug("Validating structures");
+    bool ret = true;
+
+    bool seen_ids[256] = { };
+    for (uint16_t i=0; i<num_types; i++) {
+        const struct LogStructure *logstructure = &structures[i];
+
+#if DEBUG_LOG_STRUCTURES
+        Debug("offset=%d ID=%d NAME=%s\n", i, logstructure->msg_type, logstructure->name);
+#endif
+
+        // names must be null-terminated
+        if (logstructure->name[4] != '\0') {
+            Debug("Message name not NULL-terminated");
+            ret = false;
+            validate_structures_failed();
+        }
+
+        // ensure each message ID is only used once
+        if (seen_ids[logstructure->msg_type]) {
+            Debug("ID %d used twice (LogStructure offset=%d)", logstructure->msg_type, i);
+            ret = false;
+            validate_structures_failed();
+        }
+        seen_ids[logstructure->msg_type] = true;
+
+        // ensure we have enough labels to cover columns
+        uint8_t fieldcount = strlen(logstructure->format);
+        uint8_t labelcount = count_commas(logstructure->labels)+1;
+        if (fieldcount != labelcount) {
+            Debug("fieldcount=%u does not match labelcount=%u",
+                  fieldcount, labelcount);
+            ret = false;
+            validate_structures_failed();
+        }
+
+        // check that the structure is of an appropriate length to take fields
+        const int16_t msg_len = Log_Write_calc_msg_len(logstructure->format);
+        if (msg_len != logstructure->msg_len) {
+            Debug("Calculated message length for (%s) based on format field (%s) does not match structure size (%d != %u)", logstructure->name, logstructure->format, msg_len, logstructure->msg_len);
+            validate_structures_failed();
```

Comment:
Yep, I've made it void now it is clear this function is always going to be fatal if the structures are invalid.